File: www/libs/coweb/jsoe/ContextDifference.js
//
// Difference between two contexts in terms of operations.
//
// Copyright (c) The Dojo Foundation 2011. All Rights Reserved.
// Copyright (c) IBM Corporation 2008, 2011. All Rights Reserved.
//
define([
    'coweb/jsoe/HistoryBuffer'
], function(HistoryBuffer) {
    /**
    * Stores the difference in operations between two contexts in terms of 
    * site IDs and sequence numbers.
    *
    * @ivar sites Array of integer site IDs paired with corresponding seq 
    *   numbers
    * @ivar seqs Array of integer sequence numbers paired with site IDs
    */
    var ContextDifference = function() {
        this.sites = [];
        this.seqs = [];
    };
    

    /**
     * Adds a range of operations to the difference.
     *
     * @param site Integer site ID
     * @param start First integer operation sequence number, inclusive
     * @param end Last integer operation sequence number, exclusive
     */
    ContextDifference.prototype.addRange = function(site, start, end) {
        for(var i=start; i < end; i++) {
            this.addSiteSeq(site, i);
        }
    };

    /**
     * Adds a single operation to the difference.
     *
     * @param site Integer site ID
     * @param seq Integer sequence number
     */
    ContextDifference.prototype.addSiteSeq = function(site, seq) {
        this.sites.push(site);
        this.seqs.push(seq);        
    };

    /**
     * Gets the histor buffer keys for all the operations represented in this
     * context difference.
     *
     * @return Array of keys
     */
    ContextDifference.prototype.getHistoryBufferKeys = function() {
        var arr = [];
        for(var i=0, l=this.seqs.length; i < l; i++) {
            var key = HistoryBuffer.createHistoryKey(this.sites[i], 
                this.seqs[i]);
            arr.push(key);
        }
        return arr;
    };

    /**
     * Converts the contents of this context difference to a string.
     *
     * @return String for debugging
     */
    ContextDifference.prototype.toString = function() {
        return this.getHistoryBufferKeys().toString();
    };
    
    return ContextDifference;
});

File: servers/java/coweb-operationengine/src/main/java/org/coweb/oe/ContextDifference.java
package org.coweb.oe;

import java.util.Arrays;
import java.util.Vector;

public class ContextDifference {
	
	public Vector<Integer> sites;
	public Vector<Integer> seqs;
	
	
	public ContextDifference() {
		this.sites = new Vector<Integer>();
		this.seqs = new Vector<Integer>();
	}
	
	
	/**
     * Adds a range of operations to the difference.
     *
     * @param {Number} site Integer site ID
     * @param {Number} start First integer operation sequence number, inclusive
     * @param {Number} end Last integer operation sequence number, exclusive
     */
    public void addRange(int site, int start, int end) {
        for(int i=start; i < end; i++) {
            this.addSiteSeq(site, i);
        }
    }
    
    /**
     * Adds a single operation to the difference.
     *
     * @param {Number} site Integer site ID
     * @param {Number} seq Integer sequence number
     */
    public void addSiteSeq(int site, int seq) {
        this.sites.addElement(new Integer(site));
        this.seqs.addElement(new Integer(seq));        
    }
    
    /**
     * Gets the histor buffer keys for all the operations represented in this
     * context difference.
     *
     * @return {String[]} Array of keys for HistoryBuffer lookups
     */
    public String[] getHistoryBufferKeys() {
		Vector<String> arr = new Vector<String>();
		int l = this.seqs.size();
        for(int i=0; i < l; i++) {
            String key = Operation.createHistoryKey(this.sites.elementAt(i), 
                this.seqs.elementAt(i));
            arr.addElement(key);
        }
        
        return (String[])arr.toArray();
	}
    
    /**
     * Converts the contents of this context difference to a string.
     *
     * @return {String} All keys in the difference (for debug)
     */
    public String toString() {
        return Arrays.toString(this.getHistoryBufferKeys());
    }
}
